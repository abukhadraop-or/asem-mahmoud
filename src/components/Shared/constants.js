
const NAVBAR_ITEMS = [ 'Movies', 'TV Shows', 'People', 'More'];

export const NAVBAR_SUB_ITEMS = {
  More: ['Discussions', 'Leaderboard', 'Support', 'API'],
  Movies: ['Popular', 'Now Playing', 'Upcoming', 'Top Rated'],
  People:['Popular People'],
  'TV Shows': ['Popular', 'Airing Today', 'On TV', 'Top Rated'],
};

export const APP = {
  FOOTER: {
    LIST: [
      {
        items: ['About TMDB', 'Contact Us', 'Support Forums', 'API', 'System Status'],
        name: 'The Basics',
      },
      {
        items: ['Contribution Bible', 'Add New Movie', 'Add New TV Show'],
        name: 'GET INVOLVED',
      },
      {
        items: ['Guidelines', 'Discussions', 'Leaderboard', 'Twitter'],
        name: 'COMMUNITY',
      },
      {
        items: ['Terms of Use', 'API Terms of Use', 'Privacy Policy'],
        name: 'LEGAL',
      },
    ],
    TITLE: 'Join the Community',
  },
  JOIN: 'Join TMDB',
  LANGUAGE: 'en',
  LOAD_MORE: 'Load More',
  LOGIN: 'Login',
  NAVBAR_ITEMS,
  NAVBAR_SUB_ITEMS,
  POPULAR_TITLE: 'Popular Movies',
  SIDEBAR_ITEMS: [...NAVBAR_ITEMS.slice(0, -1), ...NAVBAR_SUB_ITEMS.More ],
  SORT_FILTER: 'Sort',
  SORT_FILTERS: [
    { name: 'Popularity Descending', value: 'popularity.desc'},
    { name: 'Popularity Ascending', value: 'popularity.asc'},
    { name: 'Rating Descending', value: 'vote_average.desc'},
    { name: 'Rating Ascending', value: 'vote_average.asc'},
    { name: 'Release Date Descending', value: 'primary_release_date.desc'},
    { name: 'Release Date Ascending', value: 'primary_release_date.asc'},
    { name: 'Title (Z-A)', value: 'original_title.desc'},
    { name: 'Title (A-Z)', value: 'original_title.asc'},
  ],
  SORT_RESULT: 'Sort Results By',
};
