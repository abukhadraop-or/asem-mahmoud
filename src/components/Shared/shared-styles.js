
import styled, { createGlobalStyle } from 'styled-components';

import SourceSansPro from 'assets/fonts/SourceSansPro-Regular.ttf';
import SourceSansProBold from 'assets/fonts/SourceSansPro-Bold.ttf';
import SourceSansProLight from 'assets/fonts/SourceSansPro-Light.ttf';
import SourceConsensus from 'assets/fonts/Consensus-Bold.otf';

export const VIEW_PORT = {
  mobileView: '62.5rem',
}

export const CSSReset = createGlobalStyle`
  @font-face {
    font-family: 'Source Sans Pro';
    font-style: normal;
    font-weight: 200;
    src: url(${SourceSansPro}) format('truetype');
  }

  @font-face {
    font-family: 'Source Sans Pro Light';
    font-style: normal;
    font-weight: 200;
    src: url(${SourceSansProLight}) format('truetype');
  }

  @font-face {
    font-family: 'Consensus';
    font-style: normal;
    font-weight: 400;
    src: url(${SourceConsensus}) format('opentype');
  }

  @font-face {
    font-family: 'Source Sans Pro Bold';
    font-style: normal;
    src: url(${SourceSansProBold}) format('opentype');
  }

  :root {
    --tmdb-dark-blue: #152538;
    --tmdb-light-blue: #032541;
    --white-color: #ffffff;
    --black-color: #000;
    --primary-font: 'Source Sans Pro', Arial, sans-serif;
    --light-gray: #ffffff99;
    --dark-gray: #D3D3D3;
    --light-black: #00000026;
    --light-shade: #e3e3e3;
    --darkness-gray: #00000099;
    --dark-green: #081c22;
  }

  *,
  *::before,
  *::after {
    box-sizing: border-box;
    font-family: var(--primary-font);
    margin: 0;
    padding: 0;
  }
`;

export const Container = styled.div`
  margin: 0 auto;
  max-width: calc(100% - 1rem);
  padding: 0 0.5rem;

  @media(min-width: ${VIEW_PORT.mobileView}) {
    max-width: 83.5rem;
    padding: 0 2rem;
  }
`;

export const Items = styled.ul`
  align-items: center;
  display: flex;
  gap: 1rem;
  list-style-type: none;
`;

export const Button = styled.button`
  background: transparent;
  border: none;
  cursor: pointer;
`;
