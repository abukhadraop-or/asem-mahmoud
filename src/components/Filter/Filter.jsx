import { APP } from 'components/Shared/constants'
import React, { useContext, useState } from 'react'
import RightArrow from 'assets/images/glyphicons-basic-chevron-right.svg'
import { Button } from 'components/Shared/shared-styles'
import { FetchContext } from 'context/FetchProvider'
import {
  FilterName,
  FilterContent,
  FilterArrow,
  FilterBox,
  FilterSortName,
  FilterSortContent,
  DropDownMenu,
  DropDownItems,
  DropDownSelectedItem,
  DropDownItem,
} from './filter-styles'

export default function Filter() {
  const [toggleFilter, setToggleFilter] = useState(true)
  const [toggleDropDown, setToggleDropDown] = useState(false)
  const [selectedItem, setSelectedItem] = useState(APP.SORT_FILTERS[0].name)
  const ctx = useContext(FetchContext)

  /**
   * Handle toggle filter component.
  */
  const handleToggleFilter = () => {
    setToggleFilter(!toggleFilter)
    setToggleDropDown(false)
  }

  /**
   * Handle toggle drop down component.
   *
   * @param {string} newSelectedItem Selected item.
   * @param {string} sortBy SortBy value.
  */
    const handleToggleDropDown = (newSelectedItem, sortBy) =>  {
      setToggleDropDown(false)
      setSelectedItem(newSelectedItem)
      ctx?.setValues({ sortBy })
    }

  return (
    <div>
      <FilterBox>
        <FilterContent>
          <FilterName>{APP.SORT_FILTER}</FilterName>
          <Button onClick={handleToggleFilter}><FilterArrow rotate={toggleFilter} src={RightArrow}/></Button>
        </FilterContent>
        { toggleFilter
            && (<FilterSortContent>
              <FilterSortName>{APP.SORT_RESULT}</FilterSortName>
                <DropDownMenu>
                  <DropDownSelectedItem onClick={() => setToggleDropDown(!toggleDropDown)}>{selectedItem}</DropDownSelectedItem>
                    { toggleDropDown && <DropDownItems>
                      {APP.SORT_FILTERS.map(({ name, value }) =>
                        (name !== selectedItem && <DropDownItem key={value} onClick={() => handleToggleDropDown(name, value)}>{name}</DropDownItem>
                        ))}
                      </DropDownItems>
                    }
                </DropDownMenu>
            </FilterSortContent>)
            }
      </FilterBox>
    </div>
  )
}
