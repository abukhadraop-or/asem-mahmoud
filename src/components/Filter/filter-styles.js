import { Button, VIEW_PORT } from 'components/Shared/shared-styles';
import styled from 'styled-components';


export const FilterWrapper = styled.div``;

export const FilterBox = styled.div`
  border-radius: 0.5rem;
  border: 0.0625px solid var(--light-shade);
  box-shadow: 0 2px 8px rgb(0 0 0 / 10%);
  min-width: 100%;

  @media(min-width: ${VIEW_PORT.mobileView}) {
    min-width: 16.25rem;
  }
`;

export const FilterContent =  styled.div`
  align-items: center;
  border: 0.0625px solid var(--light-shade);
  display: flex;
  justify-content: space-between;
  padding: 0.875rem 1rem;
`;

export const FilterName = styled.div`
  font-size: 1.2rem;
  font-weight: 600;
`;

export const FilterSortContent =  styled.div`
  padding: 0.875rem 1rem;
`;

export const FilterSortName = styled.div`
  font-family: 'Source Sans Pro Light';
  margin-bottom: 0.7rem;
`;

export const FilterArrow = styled.img`
  height: 1rem;
  transform: rotate(${({ rotate }) => rotate ? 90 : 0}deg);
  width: 1rem;
`;

export const DropDownMenu = styled.div`
  position: relative;
  text-align: left;
`;

export const DropDownItem = styled(Button)`
  display: block;
  padding: 0.375rem 0.375rem 0.375rem 0.75rem;
  text-align: inherit;
  width: 100%;

  &:hover {
    background-color: var(--dark-gray);
  }
`;

export const DropDownItems = styled(FilterBox)`
  background-color: var(--white-color);
  left: 0;
  min-width: 100%;
  position: absolute;
  top: 2rem;
  z-index: 10;
`;

export const DropDownSelectedItem = styled(Button)`
  background-color: #e4e7eb;
  border-radius: 0.3rem;
  padding: 0.375rem 0 0.375rem 0.75rem;
  position: relative;
  text-align: left;
  width: 100%;

  &::before {
    position: absolute;
    content: "";
    right: 0.75rem;
    top: 40%;
    width: 0;
    height: 0;
    border: 6px solid transparent;
    border-color: var(--black-color) transparent transparent transparent;
  }
`;
