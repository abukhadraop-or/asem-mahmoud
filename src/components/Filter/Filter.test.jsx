import React from 'react'
import { shallow } from 'enzyme'
import { Button } from 'components/Shared/shared-styles'
import { APP } from 'components/Shared/constants'
import Filter from './Filter'
import '@testing-library/jest-dom'
import { DropDownItem, DropDownItems, DropDownSelectedItem, FilterArrow, FilterName, FilterSortContent } from './filter-styles'


describe('<Filter />', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(<Filter />)
  })

  it('renders correctly', () => {
    expect( wrapper).toMatchSnapshot();
  });

  it('should filter name rendered and contains correct content', () => {
    const filterName = wrapper.find(FilterName)

    expect(filterName).toHaveLength(1);
    expect(filterName.text()).toContain(APP.SORT_FILTER);
  })

  it('should filter toggle button rendered', () => {
    const toggleFilter = wrapper.find(Button);

    expect(toggleFilter).toHaveLength(1);
  })

  it('should toggle filter state true', () => {
    const toggleFilterContent = wrapper.find(FilterSortContent)
    const filterArrow = wrapper.find(FilterArrow)

    expect(filterArrow.prop('rotate')).toBeTruthy()
    expect(toggleFilterContent).toHaveLength(1)
  })

  it('should not render filter content when click on toggle filter button', () => {
    const toggleFilterButton = wrapper.find(Button)

    toggleFilterButton.simulate('click')

    expect(wrapper.find(FilterSortContent)).toHaveLength(0)
    expect(wrapper.find(FilterArrow).prop('rotate')).toBeFalsy()
  })

  it('should render dropdown items when select item', () => {
    const selectedItem = wrapper.find(DropDownSelectedItem)

    selectedItem.simulate('click')

    expect(wrapper.find(DropDownItems)).toHaveLength(1)
    expect(wrapper.find(DropDownItem)).toHaveLength(APP.SORT_FILTERS.length - 1)
  })

  it('should close dropdown items menu after select item', () => {
    const selectedItem = wrapper.find(DropDownSelectedItem)

    selectedItem.simulate('click')
    wrapper.find(DropDownItem).last().simulate('click')

    const lastItem = APP.SORT_FILTERS[APP.SORT_FILTERS.length - 1].name

    expect(wrapper.find(DropDownSelectedItem).text()).toBe(lastItem)
  })
})
