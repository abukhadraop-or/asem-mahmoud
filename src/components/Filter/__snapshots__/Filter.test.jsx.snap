// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> renders correctly 1`] = `
<div>
  <styled.div>
    <styled.div>
      <styled.div>
        Sort
      </styled.div>
      <styled.button
        onClick={[Function]}
      >
        <styled.img
          rotate={true}
          src="glyphicons-basic-chevron-right.svg"
        />
      </styled.button>
    </styled.div>
    <styled.div>
      <styled.div>
        Sort Results By
      </styled.div>
      <styled.div>
        <Styled(styled.button)
          onClick={[Function]}
        >
          Popularity Descending
        </Styled(styled.button)>
      </styled.div>
    </styled.div>
  </styled.div>
</div>
`;
